Sources/NIOCore/EventLoop.swift

```diff
@@ -1273,9 +1273,18 @@ extension EventLoopGroup {
     }
 }
 
-/// This type is intended to be used by libraries which use NIO, and offer their users either the option
+/// This type was intended to be used by libraries which use NIO, and offer their users either the option
 /// to `.share` an existing event loop group or create (and manage) a new one (`.createNew`) and let it be
 /// managed by given library and its lifecycle.
+@available(*, deprecated,
```

Comment:
I'm kinda with George here, I think it'd be better to do a multi-stage approach. Causing all user libraries to immediately get deprecation warnings without a cycle for maintainers to add new API surface seems like it's suboptimal.

Reply:
> We could just deprecate the `createNew` case. End users can provide `.shared(MultiThreadedEventLoopGroup.singleton)` until libraries add new APIs which just take `any EventLoopGroup`.

Is the suggestion that libraries like Hummingbird and friends would add their own `HummingBird.defaultEventLoopGroup` or something and migrate from `groupProvider: EventLoopGroupProvider = .createNew` to `groupProvider: EventLoopGroupProvider = .shared(HummingBird.defaultEventLoopGroup)`?

At the same time they could add a `group: any EventLoopGroup? = nil`.

That would work for me. WDYT? @Lukasa / @glbrntt & @adam-fowler

Reply:
IMO, hummingbird and co just taking an EL is not great due to the relation ship of EL to bootstrap. I think we should just create that `swift-nio-universal-bootstrap` repo at some point where we expose a common way to bootstrap a client/server/datagram with a configuration that allows the user to choose `.automatic`, `.posix` or `.niots`. With the introduction of `swift-certificates` we have a way around the TLS configuration problem since we can just accept a `Certificate` and then serialise and deserialise.

If we would have that hummingbird and friends would accept a `UniversalServerBootstrapConfiguration` which would also include the EL configuration.

Reply:
> > We could just deprecate the `createNew` case. End users can provide `.shared(MultiThreadedEventLoopGroup.singleton)` until libraries add new APIs which just take `any EventLoopGroup`.
> 
> Is the suggestion that libraries like Hummingbird and friends would add their own `HummingBird.defaultEventLoopGroup` or something and migrate from `groupProvider: EventLoopGroupProvider = .createNew` to `groupProvider: EventLoopGroupProvider = .shared(HummingBird.defaultEventLoopGroup)`?
> 
> At the same time they could add a `group: any EventLoopGroup? = nil`.
> 
> That would work for me. WDYT? @Lukasa / @glbrntt & @adam-fowler

Using the `.shared` would be the intermediate step so that end-users can avoid warnings while libraries put out new APIs.

I think the `group: any EventLoopGroup? = nil` where `nil` is "choose something sensible for me" is probably the right end goal. 

Why would libraries need to provide a `defaultEventLoopGroup`? Seems like just the `nil` case of above.

Reply:
> Why would libraries need to provide a defaultEventLoopGroup? Seems like just the nil case of above.

@glbrntt Once we deprecate `.createNew`, what should Hummingbird change its API too? https://github.com/hummingbird-project/hummingbird/blob/c5dd619b1cd0a2683f0dd3d4b7a0bea12485060b/Sources/Hummingbird/Application.swift#L68C1-L73C1

Hummingbird can't change to `eventLoopGroupProvider: NIOEventLoopGroupProvider = .shared(MultiThreadedEventLoopGroup.singleton)` because that'll switch off NIOTS support by default. And they can't use `.createNew` (because deprecated).

The only interim step that's not deprecating the whole API (which I was _originally_ intending to do with the deprecation of `NIOEventLoopGroupProvider`) is if Hummingbird provided its own `.defaultEventLoopGroup` (which switches between posix & TS) and then goes for

`eventLoopGroupProvider: NIOEventLoopGroupProvider = .shared(...defaultEventLoopGroup)`

or am I missing smth?

Alongside that, HB probably wants to introduce a new `eventLoopGroup: any EventLoopGroup? = nil` API anyway.

Reply:
> > Why would libraries need to provide a defaultEventLoopGroup? Seems like just the nil case of above.
> 
> @glbrntt Once we deprecate `.createNew`, what should Hummingbird change its API too? https://github.com/hummingbird-project/hummingbird/blob/c5dd619b1cd0a2683f0dd3d4b7a0bea12485060b/Sources/Hummingbird/Application.swift#L68C1-L73C1
> 
> Hummingbird can't change to `eventLoopGroupProvider: NIOEventLoopGroupProvider = .shared(MultiThreadedEventLoopGroup.singleton)` because that'll switch off NIOTS support by default. And they can't use `.createNew` (because deprecated).
> 
> The only interim step that's not deprecating the whole API (which I was _originally_ intending to do with the deprecation of `NIOEventLoopGroupProvider`) is if Hummingbird provided its own `.defaultEventLoopGroup` (which switches between posix & TS) and then goes for
> 
> `eventLoopGroupProvider: NIOEventLoopGroupProvider = .shared(...defaultEventLoopGroup)`
> 
> or am I missing smth?

I _think_ what you're missing is that the intermediate deprecation is to help _end users_, i.e. people using Hummingbird et. al., but not Hummingbird itself. 

What we should try to avoid is:
1. NIO is released, deprecating the provider
2. User updates and has deprecation warnings but no way to avoid them
3. Hummingbird et. al. get updated to provide an alternative API
4. User can move to alternative API

We're fine if 2 and 3 are reversed but we can't control that, so I suggest we do the following:
1. NIO is released, deprecating the `.createNew` case
2. User updates and has deprecation warning but can either live with the warning _or_ use an appropriate singleton in the `.shared` case
3. Hummingbird et. al. get updated to provide an alternative API and deprecate their API using the provider
5. User can move to new API
 
> Alongside that, HB probably wants to introduce a new `eventLoopGroup: any EventLoopGroup? = nil` API anyway.

The new API should be this ^

Reply:
> Once we deprecate .createNew, what should Hummingbird change its API too

I guess Hummingbird would deprecate the initialiser using `NIOEventLoopGroupProvider` and add a new version that took `(any EventLoopGroup)?`. Although the default parameters might complicate the issue.

Reply:
@weissi Do you have any objections to @glbrntt's proposal above?

Reply:
@Lukasa Yes, deprecating `.createNew` to start with works for me!

Reply:
I think I did the same in AHC: https://github.com/swift-server/async-http-client/blob/bdaa3b18358b60e268afe82e8feb16418874f41b/Sources/AsyncHTTPClient/HTTPClient.swift#L956-L966

Reply:
done